library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/base64/Base64Serializer.kt

```diff
@@ -0,0 +1,148 @@
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.LinkedBlockingDeque
+import java.util.concurrent.RejectedExecutionException
+import java.util.concurrent.ThreadPoolExecutor
+import java.util.concurrent.TimeUnit
+
+@Suppress("UndocumentedPublicClass")
+class Base64Serializer private constructor(
+    private val threadPoolExecutor: ExecutorService,
+    private val drawableUtils: DrawableUtils = DrawableUtils(),
+    private val base64Utils: Base64Utils = Base64Utils,
+    private val webPImageCompression: ImageCompression = WebPImageCompression()
+) {
+
+    private var asyncImageProcessingCallback: AsyncImageProcessingCallback? = null
+
+    // region internal
+
+    @MainThread
+    internal fun handleBitmap(
+        displayMetrics: DisplayMetrics,
+        drawable: Drawable,
+        imageWireframe: MobileSegment.Wireframe.ImageWireframe
+    ) {
+        asyncImageProcessingCallback?.startProcessingImage()
+
+        val bitmap = drawableUtils.createBitmapFromDrawable(drawable, displayMetrics)
+
+        if (bitmap == null) {
+            asyncImageProcessingCallback?.finishProcessingImage()
+            return
+        }
+
+        @Suppress("SwallowedException", "TooGenericExceptionCaught")
+        try {
+            threadPoolExecutor.execute {
+                val runnable = serialiseBitmap(bitmap, imageWireframe, asyncImageProcessingCallback)
+                runnable.run()
+            }
+        } catch (e: RejectedExecutionException) {
+            // TODO: REPLAY-1364 Add logs here once the sdkLogger is added
+        } catch (e: NullPointerException) {
+            // in theory will never happen but we'll log it
+            // TODO: REPLAY-1364 Add logs here once the sdkLogger is added
+        }
+    }
+
+    internal fun registerAsyncLoadingCallback(
+        asyncImageProcessingCallback: AsyncImageProcessingCallback
+    ) {
+        this.asyncImageProcessingCallback = asyncImageProcessingCallback
+    }
+
+    // endregion
+
+    // region testing
+
+    @VisibleForTesting
+    internal fun getThreadPoolExecutor(): ExecutorService = threadPoolExecutor
+
+    // endregion
+
+    // region private
+
+    private fun serialiseBitmap(
+        bitmap: Bitmap,
+        imageWireframe: MobileSegment.Wireframe.ImageWireframe,
+        asyncImageProcessingCallback: AsyncImageProcessingCallback?
+    ): Runnable = Runnable {
+        val base64String = convertBmpToBase64(bitmap)
+        finalizeRecordedDataItem(base64String, imageWireframe, asyncImageProcessingCallback)
+    }
+
+    private fun convertBmpToBase64(bitmap: Bitmap): String {
+        val byteArrayOutputStream = webPImageCompression.compressBitmapToStream(bitmap)
+
+        if (webPImageCompression.isOverSizeLimit(byteArrayOutputStream.size())) {
+            return ""
+        }
+
+        val base64 = base64Utils.serializeToBase64String(byteArrayOutputStream)
+        bitmap.recycle()
```

Comment:
I would put this in a `try/finally` to make sure you call `bitmap.recycle` even if something was thrown under the hood